rfcs/2021-10-29-9465-splunk-hec-indexer-acknowledgements.md

```diff
@@ -0,0 +1,346 @@
+
+We generate a single `guid` to use as a channel ID and include this in HEC
+requests. After submitting events to HEC, we will parse the HTTP response for an
+`ackId`. If no `ackId` is found, we rely on the current behavior of setting
+`EventStatus` based solely on the HTTP status code.
+
+If an `ackId` is found, we store it in a `Map<u64, bool>` shared with a
+background tokio task which, for all pending `ackId`’s, will query
+`/services/collector/ack`. This background task will query at an interval of 10
+seconds (as recommended by Splunk) and update the shared structure based on the
+results. Meanwhile, back in the HEC response handler, we’ll check the shared
+structure for `true` at 10 second intervals up to a maximum time limit (5
+minutes as recommended by Splunk).
+
+```{rust}
+fn call(&mut self, req: HecRequest) -> Self::Future {
+        let mut http_service = self.batch_service.clone();
+        Box::pin(async move {
+            ...
+            // handle response
+            let response = http_service.call(req).await?;
+            // if ack_id is found
+            self.ack_id_to_status_map.insert(ack_id, false);
+            let mut interval = time::interval(Duration::from_secs(10));
+            let mut retries = 0;
+            while retries < RETRY_LIMIT {
+                interval.tick().await;
+                if self.ack_id_to_status_map.get(ack_id) {
+                    // set EventStatus::Delivered and update map
+                    break;
+                }
+                retries += 1;
+            }
+
+            // if ack_id is not found, fall back on current behavior
+            let event_status = if response.status().is_success() {
+                EventStatus::Delivered
+            } else if response.status().is_server_error() {
+                EventStatus::Errored
+            } else {
+                EventStatus::Failed
+            };
+
+            ...
+        })
+    }
+```
+
+If we receive a `true` response, we remove the ackId from the shared structure
+and set `EventStatus::Delivered`.
+
+If we do not receive a `true` response and the time limit expires, we remove the
+ackId from the shared structure and set `EventStatus::Dropped`.
+
+## Alternatives
+
+* For `splunk_hec` source, [a user
+  mentioned](https://github.com/vectordotdev/vector/issues/2374#issuecomment-795367929)
+  the possibility of simply adding a mock implementation of indexer
+  acknowledgements wherein queries to `/services/collector/ack` return true for
+  any ackId. This was suggested prior to the existence of vector’s internal
+  end-to-end acknowledgements system.
+
+## Outstanding Questions
+
+* For `splunk_hec` source, there are several choices to be made in regards to
+  how closely we mimic the real Splunk HEC indexer acknowledgements behavior.
+  While we’d like to avoid inheriting Splunk’s issues wherever possible, we’d
```

Comment:
I'd agree with avoiding existing baggage - would adding additional functionality as requested by users be relatively straightforward and backward compatible?

Reply:
Good question. Most current compromises are related to Splunk channel behavior. I think this can be added later on. Channels complicate  `ackId` creation/usage but shouldn't affect the core implementation there (we'd just do it per-channel). 

Backwards compatibility is a bit tricky. If we, for example, start requiring channel IDs in requests after adding channel behavior or want to instead enable acknowledgements per authentication token, it could be hard.

Reply:
Based on the [discussion here](https://github.com/vectordotdev/vector/pull/9863#discussion_r741171896), we'll move forward with a channel based implementation. Thus, any additional functionality and backwards compatibility should be a lot simpler to support.